### PR TITLE
Introduce streamed & binary response

### PR DESCRIPTION
**Pull request by ProPheT777**
# Streamed response

```php
        /** @var \Iterator $iterator */
        $iterator = new FetcherIterator();

        $callback = function() use ($iterator) {
            $handle = fopen('php://output', 'r+');

            $columns = array('firstname', 'name');
            fputcsv($handle, $columns);

            foreach ($iterator as $r) {
                $line = array($r->getFirstname(), $r->getName());
                fputcsv($handle, $line);
            }
            fclose($handle);
        };

        Utils::makeStreamedResponse($bag, $callback, 'export.csv');
```

# Binary response

```php
        Utils::makeBinaryResponse($bag, '/path/to/a/file/to/download.pdf');
```

### CODE REVIEW

File: src/OutputBuilder/StreamedOutputBuilder.php [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+{
+    const FILENAME = 'output.streamed.file_download_name';
+
+    /** @see RFC6266 */
+    const ATTACHMENT_DISPOSITION = 'output.streamed.attachment_disposition';
+
+    public function getHeaders(ResponseBag $bag)
+    {
+        if(!$bag->has(self::FILENAME)){
+            throw new \Exception('Mission filename for streamed response');
```
Comment (cyrilf) [inline src/OutputBuilder/StreamedOutputBuilder.php]: typo: missing

Comment (oliviermadre) [inline src/OutputBuilder/StreamedOutputBuilder.php]: mission ?

File: src/Response/Utils.php [hunk 2]
```diff
@@ -0,0 +1,44 @@
+<?php
+
+namespace Pyrite\Response;
+
+use Pyrite\OutputBuilder\BinaryOutputBuilder;
+use Pyrite\OutputBuilder\StreamedOutputBuilder;
+
+class Utils
```
Comment (cyrilf) [inline src/Response/Utils.php]: Not convinced by the use of this name

File: src/Response/ResponseBagImpl.php [hunk 3]
```diff
@@ -4,18 +4,47 @@
...
+    /** @var  Callable */
+    protected $callback;
+
     public function set($key, $value)
     {
         $this->data[$key] = $value;
         return $this;
     }
 
+    public function setType($type)
```
Comment (oliviermadre) [inline src/Response/ResponseBagImpl.php]: should check the type is an available type

File: src/Stack/Application.php [hunk 4]
```diff
@@ -97,8 +100,31 @@ protected function buildResponseFromResponseBag(ResponseBag $responseBag)
         $result = $responseBag->getResult();
         $resultCode = $responseBag->getResultCode();
         $headers = $responseBag->getHeaders();
+        $type = $responseBag->getType();
 
-        return new Response($result, $resultCode, $headers);
+        if($type === ResponseBag::TYPE_DEFAULT){
```
Comment (oliviermadre) [inline src/Stack/Application.php]: should retrieve the response object through a factory, so that you don't hardcode available types in application

**Comment by cyrilf** [comment]
missing some PHPCS too, but doesn't seems to be required on this repo

File: src/OutputBuilder/StreamedOutputBuilder.php (hunk 1 above)
Comment (ProPheT777) [inline src/OutputBuilder/StreamedOutputBuilder.php]: yep typo

**Comment by cyrilf** [comment]
oh, not anymore, you're too quick :wink:

File: src/Response/Utils.php (hunk 2 above)
Comment (oliviermadre) [inline src/Response/Utils.php]: +1

File: src/Stack/Application.php (hunk 4 above)
Comment (oliviermadre) [inline src/Stack/Application.php]: up

File: src/Stack/Application.php [hunk 5]
```diff
@@ -97,8 +100,31 @@ protected function buildResponseFromResponseBag(ResponseBag $responseBag)
         $result = $responseBag->getResult();
         $resultCode = $responseBag->getResultCode();
         $headers = $responseBag->getHeaders();
+        $type = $responseBag->getType();
 
-        return new Response($result, $resultCode, $headers);
+        if ($type === ResponseBag::TYPE_DEFAULT) {
```
Comment (oliviermadre) [inline src/Stack/Application.php]: should retrieve the response object through a factory, so that you don't hardcode available types in application

File: src/OutputBuilder/StreamedOutputBuilder.php [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    /** @see RFC6266 */
+    const ATTACHMENT_DISPOSITION = 'output.streamed.attachment_disposition';
+
+    public function getHeaders(ResponseBag $bag)
+    {
+        if (!$bag->has(self::FILENAME)) {
+            throw new \Exception('Mission filename for streamed response');
+        }
+
+        $bag->addHeader('Content-Type', 'application/force-dowload');
```
Comment (vianneyb) [inline src/OutputBuilder/StreamedOutputBuilder.php]: force-download

File: src/Stack/Application.php (hunk 5 above)
Comment (ProPheT777) [inline src/Stack/Application.php]: Do we really need a factory when we don't have dependency resolution problematic (i'm not against but for 3 types and no real problematic behind)

File: src/Response/Utils.php (hunk 2 above)
Comment (ProPheT777) [inline src/Response/Utils.php]: It's recurrent naming in some OSS project for class who make easier our life but can be omitted to do it by the hand

**Comment by ProPheT777** [comment]
reminder : Security check CSV download

File: src/Response/ResponseBagImpl.php [hunk 7]
```diff
@@ -4,18 +4,51 @@
...
+    {
+        if(!in_array($type, array(self::TYPE_STREAMED, self::TYPE_BINARY, self::TYPE_DEFAULT))){
+            throw new \InvalidArgumentException('Unknown response type');
+        }
+
+        if ($type === self::TYPE_STREAMED) {
+            $this->set('format', 'streamed');
```
Comment (cyrilf) [inline src/Response/ResponseBagImpl.php]: Why is this required?

Comment (ProPheT777) [inline src/Response/ResponseBagImpl.php]: what is required ?

Comment (cyrilf) [inline src/Response/ResponseBagImpl.php]: Not required, necessary. I'm talking about setting the format to streamed here.
